keps/sig-node/00014-resource-api.md

```diff
@@ -0,0 +1,386 @@
+        - key: "Nvlink"
+          operator: "Eq"
+          values:
+            - "true"
+```
+
+Now we face the question that whether the scheduler should allow Pods requesting
+"nvidia-p100" to land on a node in this new GPU node groups. So far, we have
+received different feedbacks on this question. In some use cases, users would
+like to have minimum matching behavior that as long as the underlying hardware
+matches the minimal requirements specified through ResourceClass contraints,
+they want to allow Pods to be scheduled on the hardware. On the other hand, some users desire to reserve expensive hardware resources for users who explicitly request them.
+We feel both use cases are valid requirements. Allowing a ComputeResource to match
+multiple ResourceClasses as long as it matches their matching constraints
+perhaps yields least surprising behavior to users and also simplies upgrade
+scenario as new resource properties are introduced into the system. Therefore we
+support this behavior by default. To also provide an easy way for cluster admins
+to reserve expensive compute resources and control their access with resource
+quota, we propose to include a Priority field in ResourceClass API.
+By default, the value of this field is set to zero, but cluster admins can set
+it to a higher value, which would prevent its matching compute resources from
+being matched by lower priority ResourceClasses. i.e.,
+when a ComputeResource matches multiple ResourceClasses with different Priority values, the scheduler will choose those with the highest Priority.
+Supporting multiple ResourceClass matching also makes it easy to ensure that existing pods requesting resources through raw resource name can continue to be scheduled properly when administrators add ResourceClass in a cluster. To guarantee this, the scheduler may just consider raw resource as a special ResourceClass with empty resource metadata constraints and priority higher than any resource class.
+
+Because a ComputeResource can match multiple ResourceClasses, Scheduler and Kubelet need to ensure a consistent view on ComputeResource to ResourceClass request binding. Let us consider an example to illustrate this problem. Suppose a node has two ComputeResources, CR1 and CR2, that have the same raw resource name but different sets of properties. Suppose they both satisfy the property constraints of ResourceClass RC1, but only CR2 satisfies the property constraints of another ResourceClass RC2. Suppose a Pod requesting RC1 is scheduled first. Because the RC1 resource request can be satisfied by either CR1 or CR2, it is important for the scheduler to record the binding information and propagate it to Kubelet, and Kubelet should honor this binding instead of making its own binding decision. This way, when another Pod comes in that requests RC2, the scheduler can determine whether Pod can fit on the node or not, depending on whether the previous RC1 request is bound to CR1 or CR2.
```

Comment:
This can be used to ensure more guaranteed scheduling behavior between scheduler and kubelet for special resource scheduling requests, as scheduler now has better track on the availability of such special resources and would not over-allocate.